### ISSUE TITLE
404 NR with Istio 0.7.1

### ISSUE THREAD

**Issue by nixgadget** [istio/istio#5778]
istio 0.7.1
GKE 1.9.7-gke.1
mTLS: Disabled

With 0.7.1 we are currently unable to get service-to-service traffic going. Keep getting `404 NR` errors.

A typical error on the sending microservice proxy is,
```
[2018-05-23T11:56:37.470Z] "GET /management/health HTTP/1.1" 200 - 0 103 67 3 "-" "kube-probe/1.9+" "5bdd5af0-7ce5-9b5f-9004-bc762239b205" "10.4.8.63:9079" "127.0.0.1:9079"
[2018-05-23T11:56:38.673Z] "GET /management/health HTTP/1.1" 200 - 0 103 66 4 "-" "kube-probe/1.9+" "1ff6689c-abc4-9b64-bd6e-554a35eefa40" "10.4.8.63:9079" "127.0.0.1:9079"
[2018-05-23T11:56:39.633Z] "POST /co.riro.authorizationservice.IdentityManagementService/registerUserIdentity HTTP/2" 404 NR 0 0 0 - "-" "grpc-java-netty/1.10.0" "3e155d7e-e8bf-941b-ab3c-68af8b10b7dc" "t-auth" "-"
```

And in the sending microservice application we would also see,
```
Caused by: io.grpc.StatusRuntimeException: UNIMPLEMENTED: HTTP status code 404
invalid content-type: null
trailers: Metadata(:status=404,date=Wed, 23 May 2018 11:56:39 GMT,server=envoy)
	at io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:221)
	at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:202)
	at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:131)
	at authorizationservice.grpc.IdentityManagementServiceGrpc$IdentityManagementServiceBlockingStub.registerUserIdentity(IdentityManagementServiceGrpc.java:337)
	at onboardingservice.client.IdentityManagementServiceClient.registerUserIdentity(IdentityManagementServiceClient.java:44)
	... 84 common frames omitted
```

These services can talk to each other outside the mesh.

Any help would be good to diagnose this issue.

**Comment by nrjpoddar** [comment]
404 NR can be caused due to multiple reasons. Please check if any of the following is true in your scenario:

* No Istio routes are configured for the Host or no K8s service exists. If there is no K8s service default routing doesn't work.
* The Host header is incorrect and doesn't have the same value as the K8s service name.

**Comment by nixgadget** [comment]
@nrjpoddar I found a fix for the issue. It was due to the way i was creating the grpc channel.
I was doing,
```
NettyChannelBuilder builder = NettyChannelBuilder.forTarget()
                .nameResolverFactory(...)
                .usePlaintext();
``` 
I'm guessing `nameResolverFactory` doesnt work in the mesh for some reason.
This is completely fine outside the mesh.

**Comment by andraxylia** [comment]
This should be fixed in 0.8
